packages/arb-token-bridge-ui/src/hooks/useCompleteMultiChainTransactions.ts

```diff
@@ -0,0 +1,338 @@
+      withdrawal = await mapETHWithdrawalToL2ToL1EventResult(
+        tx,
+        parentChainProvider,
+        chainProvider,
+        tx.parentChainId,
+        tx.chainId
+      )
+    }
+  }
+
+  if (withdrawal) {
+    return transformWithdrawal(withdrawal)
+  }
+}
+
+function getProvider(chainId: ChainId) {
+  const rpcUrl =
+    rpcURLs[chainId] ?? getWagmiChain(chainId).rpcUrls.default.http[0]
+  return new StaticJsonRpcProvider(rpcUrl)
+}
+
+/**
+ * Fetches transaction history only, and only for a specific direction.
+ * The query could be e.g. deposits or withdrawals.
+ */
+const useTransactionListByDirection = (
+  direction: 'deposits' | 'withdrawals'
+) => {
+  const [transactions, setTransactions] = useState<DepositOrWithdrawal[][]>([])
+  const [page, setPage] = useState(0)
+  const [loading, setLoading] = useState(true)
+
+  const latestTransactions = useLatest(transactions)
+  const latestPage = useLatest(page)
+
+  const { address } = useAccount()
+
+  const shouldFetchPage = useCallback(
+    (index: number) => {
+      const _page = latestPage.current
+
+      if (_page === 0) {
+        return true
+      }
+
+      const txCount = latestTransactions.current[index]?.length ?? 0
+      // fetch next page for a chain pair if all fetched pages are full
+      // we check for >= in case some txs were included by initiating a transfer
+      return txCount / _page >= PAGE_SIZE
+    },
+    [latestTransactions, latestPage]
+  )
+
+  const { data, error } = useSWRImmutable(
+    address ? [direction, address, page] : null,
+    ([_direction, _address, _page]) => {
+      return Promise.all(
+        multiChainFetchList.map((c, index) => {
+          if (!shouldFetchPage(index)) {
+            return []
+          }
+
+          const params = {
+            sender: _address,
+            receiver: _address,
+            l1Provider: getProvider(c.parentChain),
+            l2Provider: getProvider(c.chain),
+            pageNumber: _page,
+            pageSize: PAGE_SIZE
+          }
+
+          return _direction === 'deposits'
+            ? fetchDepositList(params)
+            : fetchWithdrawalList(params)
+        })
+      )
+    }
+  )
+
+  useEffect(() => {
+    if (data) {
+      // if there is not a single full page in any of the data fetched, then there is no more to fetch
+      const shouldFetchNextPage = data.some((_, index) =>
+        shouldFetchPage(index)
+      )
+
+      // include the new data with the previously fetched data
+      // the data is grouped by chain pairs
+      setTransactions(prevTransactions => {
+        return data.map((transactionsForChainPair, chainPairIndex) => {
+          const prevTransactionsForChainPair = prevTransactions[chainPairIndex]
+          return [
+            ...(prevTransactionsForChainPair
+              ? prevTransactionsForChainPair
+              : []),
+            ...(transactionsForChainPair ? transactionsForChainPair : [])
+          ]
+        })
+      })
```

Comment:
for readability, i think it's better to do this
```suggestion
      setTransactions(prevTransactions =>
        data.map((transactionsForChainPair, chainPairIndex) => [
          ...(prevTransactions[chainPairIndex] ?? []),
          ...(transactionsForChainPair ?? [])
        ])
      )
```

Reply:
nice, done